### PR TITLE
Pod os field kubelet

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Once #104613 and #104693
merge, we'll have OS field in pod spec. Kubelet will also start reconciling labels.
From then on, Kubelet should start rejecting pods
where pod.Spec.OS and node's OS label won't match
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Kubelet should reject pods whose OS doesn't match the node's OS label.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
[KEP]: https://github.com/kubernetes/enhancements/pull/2803
```

### CODE REVIEW

**Comment by ehashman** [comment]
What is the KEP for this?

/label api-review

**Comment by ravisantoshgudimetla** [comment]
/remove label api-review

@ehashman - I updated the description with KEP associated. The api review can be done on https://github.com/kubernetes/kubernetes/pull/104693. I picked the commits from that PR to make the testing easier.

File: pkg/kubelet/lifecycle/predicate.go
Comment (derekwaynecarr) [inline pkg/kubelet/lifecycle/predicate.go:161]: nit: Mismatch and not MisMatch

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
one minor nit , this looks generally good after the prereq pr merges.

**Comment by ravisantoshgudimetla** [comment]
/hold 

till https://github.com/kubernetes/kubernetes/pull/104613 merges

**Comment by derekwaynecarr** [comment]
/lgtm
/approve

remove the hold after the requisite pr merges (just tagged)

File: pkg/kubelet/lifecycle/predicate.go
Comment (liggitt) [inline pkg/kubelet/lifecycle/predicate.go:190]: don't compare the pod os to the node labels, compare it to runtime.GOOS

Comment (liggitt) [reply pkg/kubelet/lifecycle/predicate.go:190]: the pod nodeSelector should be compared to the node labels with the OS/Arch label ensured to match the real runtime.GOOS/GOARCH, but that's distinct from checking the pod.spec.os against runtime.GOOS

**Comment by ravisantoshgudimetla** [comment]
@liggitt suggested to not rely on featuregate in kubelet to reject pods. He also suggested to add another featuregate for Pod security standards so that they can get merged in this release instead of [waiting till ](https://github.com/kubernetes/kubernetes/pull/105919#issuecomment-952291110)1.26 or 1.27(when `IdentifyPodOS` feature graduates to GA).

File: pkg/kubelet/lifecycle/predicate.go
Comment (liggitt) [inline pkg/kubelet/lifecycle/predicate.go:168]: drop NodeLabel from the reason… we're not comparing to the label

Comment (liggitt) [reply pkg/kubelet/lifecycle/predicate.go:168]: pick a reason that will continue to make sense if LCOW is ever supported… maybe `PodOSNotSupported` or something?

Comment (ravisantoshgudimetla) [reply pkg/kubelet/lifecycle/predicate.go:168]: The only reason for keeping it is - the user may not know about the OS of the node. Sometimes the only object he/she may have access to is node object.

**Pull request merged**
